oak_sign/src/main.rs

```diff
@@ -0,0 +1,197 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! An utility binary to sign Oak Wasm modules.
+//!
+//! Invoke with:
+//!
+//! ```shell
+//! cargo run --manifest-path=oak_sign/Cargo.toml -- \
+//!     generate \
+//!     --private-key=<PRIVATE_KEY_FILE>.key \
+//!     --public-key=<PUBLIC_KEY_FILE>.pem
+//!
+//! cargo run --manifest-path=oak_sign/Cargo.toml -- \
+//!     sign \
+//!     --private-key=<PRIVATE_KEY_FILE>.key \
+//!     --module=<WASM_MODULE_FILE>.wasm \
+//!     --signature=<SIGNATURE_FILE>.sign
+//!
+//! cargo run --manifest-path=oak_sign/Cargo.toml -- \
+//!     verify \
+//!     --module=<WASM_MODULE_FILE>.wasm \
+//!     --signature=<SIGNATURE_FILE>.sign
+//! ```
+
+use anyhow::Context;
+use log::info;
+use pem::{encode, encode_many, parse, parse_many, Pem};
+use ring::{
+    rand,
+    signature::{self, KeyPair},
+};
+use std::{
+    collections::HashMap,
+    fs::{read, write},
+};
+use structopt::StructOpt;
+
+/// Command line options for the Oak sign.
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Oak Sign")]
+struct Opt {
+    #[structopt(subcommand)]
+    cmd: Command,
+}
+
+/// Available commands for the Oak sign.
+#[derive(StructOpt, Clone)]
+enum Command {
+    Generate(Generate),
+    Sign(Sign),
+    Verify(Verify),
+}
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Generate Ed25519 key pair")]
+struct Generate {
+    #[structopt(long, help = "Output private key file path")]
+    pub private_key: String,
+    #[structopt(long, help = "Output public key file path")]
+    pub public_key: String,
+}
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Sign Wasm module")]
+struct Sign {
+    #[structopt(long, help = "Input private key file path")]
+    pub private_key: String,
+    #[structopt(long, help = "Input Wasm module path")]
+    pub module: String,
+    #[structopt(long, help = "Output signature path")]
+    pub signature: String,
+}
+
+#[derive(StructOpt, Clone)]
+#[structopt(about = "Verify Wasm module signature")]
+struct Verify {
+    #[structopt(long, help = "Input signature path")]
+    pub signature: String,
+    #[structopt(long, help = "Input Wasm module path")]
+    pub module: String,
+}
+
+// PEM file tags.
+const PRIVATE_KEY_TAG: &str = "PRIVATE KEY";
+const PUBLIC_KEY_TAG: &str = "PUBLIC KEY";
+const SIGNATURE_TAG: &str = "SIGNATURE";
+
+/// Creates a PEM structure for encoding.
+fn create_pem(tag: &str, contents: &[u8]) -> Pem {
+    Pem {
+        tag: tag.to_string(),
+        contents: contents.to_vec(),
+    }
+}
```

Comment:
I'm curious why you decided to use PEM for storing this output, could you explain how you got there? I imagine other options may be protobuf or TOML or JSON, curious how they compare.

(I'm not suggesting changing to any of those format, I would just like to learn more about the differences myself)

Reply:
I was thinking of PEM as a key format that everyone is familiar with (and got used to).
Also PEM implies using base64 for encoding keys, which would not be obvious in TOML or JSON without an additional explanation.

Reply:
Thus, using non-standard formats would create similar problem as described in https://github.com/project-oak/oak/issues/1385:
- Someone, who would want to check the signature would need to go deep in the documentation in order to understand, in what encoding keys and signatures are stored.